internal/v4/api_test.go

```diff
@@ -885,6 +917,85 @@ func (s *APISuite) TestMetaPerm(c *gc.C) {
 			Write: []string{},
 		},
 	})
+
+	// Restore some write rights to the stable channel.
+	s.assertPutAsAdmin(c, "trusty/wordpress-1/meta/perm/write", []string{"bob"})
+
+	// ~charmers/trusty/wordpress-1 has been published only to the
+	// stable channel. If we specify a channel in a perm put
+	// request, we'll get an error because the channel isn't valid for
+	// that entity.
+	s.doAsUser("charmers", func() {
+		httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+			Handler: s.srv,
+			Do:      bakeryDo(nil),
+			Method:  "PUT",
+			JSONBody: params.PermRequest{
+				Read:  []string{"foo"},
+				Write: []string{"bar"},
+			},
+			URL:          storeURL("trusty/wordpress-1/meta/perm?channel=development"),
+			ExpectStatus: http.StatusNotFound,
+			ExpectBody: params.Error{
+				Code:    params.ErrNotFound,
+				Message: `cs:trusty/wordpress-1 not found in development channel`,
+			},
+		})
+	})
+
+	// Similarly, we should be able to specify a channel on read
+	// to read a different channel.
+	s.doAsUser("bob", func() {
+		s.assertGet(c, "trusty/wordpress/meta/perm?channel=unpublished", params.PermResponse{
+			Read:  []string{"bob"},
+			Write: []string{"admin"},
+		})
+		s.assertGet(c, "wordpress/meta/perm?channel=development", params.PermResponse{
+			Read:  []string{"bob", "charlie"},
+			Write: []string{"charmers"},
+		})
+	})
+
+	// We can't write to a channel that the charm's not in.
```

Comment:
I think this should be allowed. It's not a fundemental part of the design, but an artifact of the implementation that disallows this.